File: partiql-ast/src/main/java/org/partiql/ast/v1/CreateTable.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public class CreateTable extends DDL {
    @NotNull
    public Identifier name;

    @Nullable
    public TableDefinition definition;

    public CreateTable(@NotNull Identifier name, @Nullable TableDefinition definition) {
        this.name = name;
        this.definition = definition;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(name);
        if (definition != null) {
            kids.add(definition);
        }
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitCreateTable(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/CreateTable.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class CreateTable(
    @JvmField
    public var name: Identifier,
    @JvmField
    public var definition: TableDefinition?,
) : DDL() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(name)
        definition?.let { kids.add(it) }
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitCreateTable(this, ctx)
}
